pkg/services/ngalert/state/manager.go

```diff
@@ -58,13 +63,13 @@ type ManagerCfg struct {
 	Historian     Historian
 	// DoNotSaveNormalState controls whether eval.Normal state is persisted to the database and returned by get methods
 	DoNotSaveNormalState bool
-	// MaxStateSaveConcurrency controls the number of goroutines (per rule) that can save alert state in parallel.
-	MaxStateSaveConcurrency int
 
 	// ApplyNoDataAndErrorToAllStates makes state manager to apply exceptional results (NoData and Error)
 	// to all states when corresponding execution in the rule definition is set to either `Alerting` or `OK`
 	ApplyNoDataAndErrorToAllStates bool
 
+	StatePersister StatePersister
```

Comment:
Could we avoid putting actual dependencies, (i.e., things that are not configuration) onto the configuration struct? This causes all kinds of cyclic problems when you want to configure the elements that are also members of this struct. We could instead pass the persister to the manager as a separate arg.

I realize `Historian` also does this, but I think it's bad there too. It would be nice if `Cfg` struct was a bundle of static configuration.

Then, you initialize the manager like `NewManager(cfg, statePersister)`?

Reply:
Very good catch, wasn't thinking about this.